### PR TITLE
`metatensor-learn` Dataset and Dataloader

### PR DESCRIPTION
**Pull request by jwa7**
A `Dataset`, `IndexedDataset`, `DataLoader`, and two collate functions for integrating `metatensor` with `torch`

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [ ] Issue referenced (for PRs that solve an issue)? n/a

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
:books: Documentation preview :books:: https://metatensor--428.org.readthedocs.build/en/428/

### CODE REVIEW

File: python/metatensor-learn/metatensor/learn/data/dataset.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    """
+    Class for defining a torch dataset.
+
+    `data` must be a dict of data items indexed by their str name. Each data
+    item is either a list or a dict.
+    """
+
+    def __init__(
+        self,
+        data: Dict[str, Union[List, Dict]],
```
Comment (agoscinski) [inline python/metatensor-learn/metatensor/learn/data/dataset.py:19]: Sorry for forgetting, what was the use case of using a `Dict[str, Dict]` again? If you want to give your samples more meta information, like structure, so we are agnostic to the data format used (TensorMap atomistic.System)?

File: python/metatensor-learn/metatensor/learn/data/collate.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+"""
+Module containing a collate function for use in a Dataloader.
+"""
+
+import metatensor
+
+
+def collate_fn(batch, join: bool = False):
```
Comment (agoscinski) [inline python/metatensor-learn/metatensor/learn/data/collate.py:8]: So this would be the TensorMap collate? and for other data types we would have a different one?

Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/collate.py:8]: Good point. I suppose if the underlying data were `Systems` objects, for instance, the above collate function would work with `join=False`? I think I might have missed this in the discussion, but what are the other ways that one would want to collate samples that aren't TensorMaps?

File: python/metatensor-learn/metatensor/learn/data/dataset.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                    # later - do nothing yet
+                    self._data[name][sample_idx] = item[sample_idx]
+
+    def __len__(self) -> int:
+        """
+        Returns the length (i.e. number of samples) of the dataset
+        """
+        return len(self._sample_idxs)
+
+    def __getitem__(self, sample_idx: int, names: Optional[List[str]] = None) -> List:
```
[the remark this answers is not in the document]
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/dataset.py:106]: Yeah so I'm wondering whether this should return a dict or a list. It could be helpful to return a dict so that you have access to the sample indices, which is useful when getting a batch from a dataloader for instance. Thoughts?

File: python/metatensor-learn/metatensor/learn/data/collate.py (hunk 2 above)
Comment (abmazitov) [reply python/metatensor-learn/metatensor/learn/data/collate.py:8]: I think the general idea is the following: in the input is a TensorMap, or a torch.Tensor, we can join it (or not). For the rest in the inputs, we just return a subset of the original set of data (i.e. a sub-list or sub-dict).

It may be implemented in a following way (inspired by torch_geometric Dataloader)

```py
class DataLoader(torch.utils.data.DataLoader):
    def __init__(
        self,
        dataset: Union[Dataset, Sequence[BaseData]],
        batch_size: int = 1,
        shuffle: bool = False,
        **kwargs,
    ):

        self.collator = Collater(dataset) # A class for determining the behavior of the collate_fn, may be actually replaced with a single collate_fn, covering all use cases

        if isinstance(dataset, OnDiskDataset):
            dataset = range(len(dataset)) # in this case we do not collate the objects

        super().__init__(
            dataset,
            batch_size,
            shuffle,
            collate_fn=self.collator.collate_fn,
            **kwargs,
        )
```

So if the data type is recognised (i.e. it is a TensorMap, or a torch.Tensor the Collater will join the objects either by `metatensor.join()` or `torch.stack() / torch.cat()`). Otherwise it's just a list or dict of objects.

File: python/metatensor-learn/metatensor/learn/data/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    def __init__(
+        self, dataset, collate_fn: Callable, batch_size: Optional[int] = None, **kwargs
+    ):
+        self._sample_idxs = dataset._sample_idxs
+        self._collate_fn = collate_fn
+        self._batch_size = (
+            batch_size if batch_size is not None else len(self._sample_idxs)
+        )
+
+        self._dataloader = torch.utils.data.DataLoader(
```
[the remark this answers is not in the document]
Comment (abmazitov) [reply python/metatensor-learn/metatensor/learn/data/dataloader.py:23]: Agree, we should inherit from torch Dataloader

File: python/metatensor-learn/metatensor/learn/data/dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abmazitov) [reply python/metatensor-learn/metatensor/learn/data/dataset.py:106]: Is this the `OnDiskDataset` or you want to create a single Dataset for all the use cases? Because as far as I remember, the idea for to create a single Dataloader, but two datasets: `InMemoryDataset` and `OnDiskDataset`.

File: python/metatensor-learn/metatensor/learn/data/collate.py (hunk 2 above)
Comment (agoscinski) [reply python/metatensor-learn/metatensor/learn/data/collate.py:8]: Another question I have comes from looking at this example https://pytorch.org/tutorials/beginner/data_loading_tutorial.html . In this example they work with a dict dataset, and return the batches in the same dict format. While we do not need to replicate this necessary, it seems to me more intuitive. So in your notebook example would be
```python
batch = next(iter(train_dataloader)) # a dict with keys "descriptor", "target", "auxillary"
in_train_batch, out_train_batch, aux_train_batch = batch.values()
```
So here I get the same format out as I put in the data in the Dataset.

File: python/metatensor-learn/metatensor/learn/data/dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/dataset.py:91]: Ok so I've changed it to return the sample indices as part of the batch, but fixed the type to be a List that is easily zip-able by the collate_fn. I've found this particularly useful when joining TensorMaps, as you can easily know the batch size (i.e. when the dataset isn't whole multiple of `batch_size` and there are remainders, for instance) and indices in that batch.

File: python/metatensor-learn/metatensor/learn/data/dataset.py (hunk 1 above)
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/dataset.py]: To be fair, I think that there isn't a use case for this. I have removed this, and required that data is passed as `Dict[str, List]`. A non-continuous range of indices can be passed in `sample_idxs`, that will be zipped 1:1 with the elements in the data items

File: python/metatensor-learn/metatensor/learn/data/dataloader.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    """
+    Class for loading data from the Dataset class.
+    """
+
+    def __init__(
+        self, dataset, collate_fn: Callable, batch_size: Optional[int] = None, **kwargs
+    ):
+        self._sample_idxs = dataset._sample_idxs
+        self._collate_fn = collate_fn
+        self._batch_size = (
```
[the remark this answers is not in the document]
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/dataloader.py]: Good point. I've removed everything from dataloader as arguments to `__init__` except the dataset and collate_fn. I think this is better as the user can pass whatever kwargs they like, as long as they are accepted by torch.Dataset. In this case the `batch_size` default will be inherited from the torch Dataset class

File: python/metatensor-learn/metatensor/learn/data/dataloader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/dataloader.py:21]: Good spot! Changed :)

File: python/metatensor-learn/metatensor/learn/data/collate.py (hunk 2 above)
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/collate.py]: So I've implemented 2 collate functions now, see comment below.

**Comment by jwa7** [comment]
Summary of the changes in the last commit:

- Dataset now requires `data` arg is passed as `Dict[str, List]`. If `sample_idxs` is not passed, a continuous range [0, ..., n_samples] is used, otherwise the elements of `sample_idxs` is zipped with the elements in each data item. Internally, the samples for each item are stored as a dict, indexed by their sample index.
- `__getitem__` of the Dataset now returns a list, where the first element in the list is the sample index.
- I've made `collate_fn` a generic collate for arbitrary objects. It only zips the list of lists returned from the dataset. I've made `collate_tensormaps` a more specific collate fn, where the zipping is followed by joining (or not, if `join=False`) of the TensorMap data items.
- I've stripped back the DataLoader to only explicitly ask for args `dataset` and `collate_fn`. This is more flexible as the user can pass `kwargs` for anything that the native torch dataset accepts - `batch_size`, `shuffle`, `sampler`, etc, with the defaults inherited.

File: python/metatensor-learn/metatensor/learn/data/dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abmazitov) [reply python/metatensor-learn/metatensor/learn/data/dataset.py:91]: Yeah, it's ok, but my question was - is it a prototype of the OnDiskDataset or a Dataset in general? Because, let say, if we have an input as a List[System] or Dict[str, System], it already seems to go out of the scope of the data types you assigned with the input.

File: python/metatensor-learn/metatensor/learn/data/collate.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
+"""
+Module containing a collate function for use in a Dataloader.
+"""
+from typing import List
+import metatensor
+from metatensor import TensorMap
+
+
+def collate_fn(batch: List[List]) -> List[List]:
```
Comment (agoscinski) [inline python/metatensor-learn/metatensor/learn/data/collate.py:9]: Sorry if I misunderstood, but I had the feeling this point was not addressed in the last discussion because it got intermingled with another point. I am okay if we don't do it, but I had the feeling it was not even mentioned

Another question I have comes from looking at this example https://pytorch.org/tutorials/beginner/data_loading_tutorial.html . In this example they work with a dict dataset, and return the batches in the same dict format. While we do not need to replicate this necessary, it seems to me more intuitive. So in your notebook example would be
```python
batch = next(iter(train_dataloader)) # a dict with keys "descriptor", "target", "auxillary"
in_train_batch, out_train_batch, aux_train_batch = batch.values()
```
So here I get the same format out as I put in the data in the Dataset.

https://github.com/lab-cosmo/metatensor/pull/428#discussion_r1417058611

This would also affect the  also the `__get_item__` function in Dataset. So it would return a dict
```python
data_i = dataset[i] -> a dict with keys "descriptor", "target", "auxillary" and inside there is again a list or another dict

```

File: python/metatensor-learn/metatensor/learn/data/dataloader.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (agoscinski) [reply python/metatensor-learn/metatensor/learn/data/dataloader.py:21]: Now you inherit from dataloader **and** have dataloader as member variable. With 
```python
from  .collate import collate_fn
class DataLoader(torch.utils.data.DataLoader):
    """
    Class for loading data from an `:py:class:`torch.utils.data.Dataset` object with a
    default collate function that supports :py:class:`torch.Tensor`,
    :py:class:`atomistic.Systems`, or :py:class:`TensorMap`. Any argument as accepted by
    the native :py:class:`torch.utils.data.DataLoader` class is supported.  Please refer
    to https://pytorch.org/docs/stable/data.html
    """

    def __init__(self, dataset: torch.utils.data.Dataset, collate_fn: Callable = collate_fn, **kwargs):
        super().__init__(dataset, collate_fn=collate_fn, **kwargs)
```
you should be able to omit the member variable and also don't need to redefine `__iter__`. Our Dataloader is only a thin wrapper using our `collate_fn` function. The doc is just a suggestion.

[the remark this answers is not in the document]
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/dataloader.py]: Nice thanks! :)

File: python/metatensor-learn/metatensor/learn/data/dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/dataset.py]: Resolving as now outdated - using `kwargs` instead. But yes, by passing callables we are providing a more general Dataset where the user can lazily load from disk some or all of the data

File: python/metatensor-learn/metatensor/learn/data/dataloader.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+"""
+Module containing the DataLoader.
+"""
+from typing import Callable
+import torch
+
+from collate import collate_fn
```
Comment (Luthaf) [inline python/metatensor-learn/metatensor/learn/data/dataloader.py:7]:
```suggestion
from .collate import collate_fn
```

File: python/metatensor-learn/metatensor/learn/data/collate.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+"""
+from collections import namedtuple
+from typing import List, Tuple
+
+import torch
+import metatensor
+from metatensor import TensorMap
+
+
+def collate_fn(batch: List[Tuple]) -> Tuple:
```
Comment (Luthaf) [inline python/metatensor-learn/metatensor/learn/data/collate.py:12]: As I understand the changes here, we don't support mixed data types in the dataset, right? 

This would not be able to handle something like `Dataset(frames=[system, system], energy=[TensorMap(), TensorMap()])`? I think we want to have something like this.

File: python/metatensor-learn/tests/dataloader.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+import metatensor
+from metatensor.learn.data import DataLoader
+
+from .dataset import dataset_in_mem, dataset_mixed_mem_disk, dataset_on_disk
+from .utils import SAMPLE_INDICES
+
+
+@pytest.mark.parametrize(
+    "dset", [dataset_in_mem, dataset_on_disk, dataset_mixed_mem_disk]
```
Comment (Luthaf) [inline python/metatensor-learn/tests/dataloader.py:15]:
```suggestion
    "create_dataset", [dataset_in_mem, dataset_on_disk, dataset_mixed_mem_disk]
```

File: python/metatensor-learn/tests/data/water.xyz
Comment (Luthaf) [inline python/metatensor-learn/tests/data/water.xyz:1]: We should not have such large file in the repo. Why do we need this (compared to creating 100 random frames with ase at runtime?)

File: python/metatensor-learn/metatensor/learn/data/collate.py (hunk 8 above)
Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/collate.py]: No, this is supported! You can pass any arbitrary object in each of the fields

File: python/metatensor-learn/metatensor/learn/data/dataloader.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader
+    """
+
+    def __init__(
+        self,
+        dataset: torch.utils.data.Dataset,
+        collate_fn: Callable = collate_fn,
+        **kwargs,
+    ):
+        super().__init__(dataset, collate_fn=group, **kwargs)
```
Comment (abmazitov) [inline python/metatensor-learn/metatensor/learn/data/dataloader.py:28]: Shouldn't we use `collate_fn=collate_fn` in the `super(). __init__()`?

Comment (jwa7) [reply python/metatensor-learn/metatensor/learn/data/dataloader.py]: Woops yes thanks! Just updated

File: python/metatensor-learn/metatensor/learn/data/collate.py [hunk 11]
```diff
@@ -35,7 +35,11 @@ def group(batch: List[Tuple]) -> Tuple:
     return namedtuple("Batch", names)(*list(zip(*batch)))
 
 
-def group_and_join(batch: List[Tuple]) -> Tuple:
+def group_and_join(
+    batch: List[Tuple],
+    fields_to_join: Optional[List[str]] = None,
+    join_kwargs: Optional[dict] = {},
```
Comment (Luthaf) [inline python/metatensor-learn/metatensor/learn/data/collate.py]:
```suggestion
    join_kwargs: Optional[dict] = None,
```

If you have optional, this feels more natural. Also, you should never use mutable values (i.e. empty dict/list) as default arguments: https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument

**Comment by Luthaf** [comment]
Now that #426 has been merged, can you rebase this PR on top of master?

**Comment by jwa7** [comment]
Think we're good for review! All that's left to do is a rebase once #447 is merged, to fit with the new structure of the docs

**Comment by Luthaf** [comment]
I did a pass over the tutorials, please have a look at the latest commit to check the changes. I still have to finish a review of the code itself!

**Review by Luthaf — COMMENTED** [review]
Ok, the code looks good. There is one small issue with the collate function which does not support `metaensor.torch.TensorMap`, and an even smaller issue that the tutorials all have the same thumbnail in https://metatensor--428.org.readthedocs.build/en/428/examples/learn/index.html

**Review by Luthaf — APPROVED** [review]
